x/perp/v2/keeper/liquidate.go

```diff
@@ -40,12 +55,75 @@ func (k Keeper) MultiLiquidate(
 	}
 
 	if allFailed {
-		return resp, types.ErrAllLiquidationsFailed.Wrapf("%d liquidations failed", len(liquidationRequests))
+		prettyResps, errPrettyResp := PrettyLiquidateResponse(resp)
+
+		numLiquidations := len(liquidationRequests)
+		errDescription := strings.Join(
+			[]string{
+				fmt.Sprintf("%d liquidations failed", numLiquidations),
+				fmt.Sprintf("liquidate_responses: %s", prettyResps),
+			},
+			"\n",
+		)
+		if errPrettyResp != nil {
+			errDescription += fmt.Sprintf("\n%s", errPrettyResp.Error())
+		}
+		return resp, types.ErrAllLiquidationsFailed.Wrap(errDescription)
 	}
 
 	return resp, nil
 }
 
+/*
+PrettyLiquidateResponse converts a slice of liquidation responses into a
+pretty formatted JSON array for each response. This helps with providing
+descriptive error messages in the case of failed liquidations.
+
+Example outputs:
+
+```json
+[
+
+	{
+	  "success": false,
+	  "error": "failed liquidation A",
+	  "liquidator_fee": null,
+	  "perp_ef_fee": null,
+	  "trader": "dummytraderA"
+	},
+	{
+	  "success": true,
+	  "error": "",
+	  "liquidator_fee": { denom: "unibi", amount: "420"},
+	  "perp_ef_fee": null,
+	  "trader": "dummytraderB"
+	}
+
+]
+```
+*/
+func PrettyLiquidateResponse(
```

Comment:
It looks like a lot is happening here, but it's just converting each `proto.Message` for the liquidate responses into a JSON string.